File: .agents/skills/imagegen-codex/scripts/remove_chroma_key.js
#!/usr/bin/env node
/**
 * Remove a solid chroma-key background from an image using sharp.
 * Supports imagegen skill transparent workflow in Node.js environment.
 */

const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

function parseArgs() {
  const args = process.argv.slice(2);
  const options = {
    input: '',
    out: '',
    key: '#00ff00',
    tolerance: 100
  };

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '-i' || args[i] === '--input') {
      options.input = args[++i];
    } else if (args[i] === '-o' || args[i] === '--out') {
      options.out = args[++i];
    } else if (args[i] === '-k' || args[i] === '--key') {
      options.key = args[++i];
    } else if (args[i] === '-t' || args[i] === '--tolerance') {
      options.tolerance = parseInt(args[++i], 10);
    }
  }

  return options;
}

async function main() {
  const options = parseArgs();

  if (!options.input || !options.out) {
    console.error('Usage: node remove_chroma_key.js -i <input_image> -o <output_image> [-k <key_color_hex>] [-t <tolerance>]');
    process.exit(1);
  }

  const inputPath = path.resolve(options.input);
  const outputPath = path.resolve(options.out);

  if (!fs.existsSync(inputPath)) {
    console.error(`Error: Input file does not exist: ${inputPath}`);
    process.exit(1);
  }

  // Ensure output directory exists
  const outDir = path.dirname(outputPath);
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true });
  }

  // Parse key color
  const hex = options.key.replace('#', '');
  if (hex.length !== 6) {
    console.error('Error: Key color must be a 6-character hex string (e.g. #00ff00)');
    process.exit(1);
  }

  const keyR = parseInt(hex.substring(0, 2), 16);
  const keyG = parseInt(hex.substring(2, 4), 16);
  const keyB = parseInt(hex.substring(4, 6), 16);
  const tolerance = options.tolerance;

  console.log(`Processing image: ${inputPath}`);
  console.log(`Removing key color R:${keyR} G:${keyG} B:${keyB} with tolerance: ${tolerance}...`);

  try {
    const image = sharp(inputPath);
    
    // Get raw RGBA pixel data
    const { data, info } = await image
      .ensureAlpha()
      .raw()
      .toBuffer({ resolveWithObject: true });

    // Process pixels and set alpha to 0 for key color matches
    for (let i = 0; i < data.length; i += 4) {
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];

      // Calculate Euclidean distance in RGB space
      const dist = Math.sqrt((r - keyR) ** 2 + (g - keyG) ** 2 + (b - keyB) ** 2);

      if (dist < tolerance) {
        data[i + 3] = 0; // Fully transparent
      } else if (dist < tolerance + 30) {
        // Soft edge anti-aliased transition
        const factor = (dist - tolerance) / 30.0;
        data[i + 3] = Math.floor(255 * factor);
      }
    }

    // Save processed pixel buffer to destination files (support WebP auto-conversion)
    const isWebpOutput = outputPath.endsWith('.webp');
    const webpPath = isWebpOutput ? outputPath : outputPath.replace(/\.[^/.]+$/, "") + ".webp";

    const sharpInstance = sharp(data, {
      raw: {
        width: info.width,
        height: info.height,
        channels: 4
      }
    });

    if (isWebpOutput) {
      await sharpInstance.webp({ quality: 85 }).toFile(outputPath);
      console.log(`Success! Transparent WebP image saved to: ${outputPath}`);
    } else {
      // Save PNG first
      await sharpInstance.clone().png().toFile(outputPath);
      console.log(`Success! Transparent PNG image saved to: ${outputPath}`);

      // Auto-convert to WebP
      await sharpInstance.webp({ quality: 85 }).toFile(webpPath);
      console.log(`Auto-converted! Transparent WebP version saved to: ${webpPath}`);
    }
  } catch (error) {
    console.error('Error processing image:', error);
    process.exit(1);
  }
}

main();

File: .agents/skills/imagegen-codex/scripts/remove_chroma_key.py
#!/usr/bin/env python3
"""Remove a solid chroma-key background from an image.

This helper supports the imagegen skill's transparent workflow:
generate an image on a flat key color, then convert that key color to alpha.
"""

import argparse
from pathlib import Path
import sys

def main():
    parser = argparse.ArgumentParser(description="Remove chroma-key background from an image.")
    parser.add_argument("-i", "--input", required=True, help="Path to input image")
    parser.add_argument("-o", "--out", required=True, help="Path to output image (PNG or WebP)")
    parser.add_argument("-k", "--key", default="#00ff00", help="Chroma key color in hex (default: #00ff00)")
    parser.add_argument("-t", "--tolerance", type=int, default=100, help="Color distance tolerance (0-255)")
    parser.add_argument("-f", "--force", action="store_true", help="Force overwrite output if exists")
    
    args = parser.parse_args()
    
    try:
        from PIL import Image
    except ImportError:
        print("Error: Pillow is required. Install it using 'pip install pillow'", file=sys.stderr)
        sys.exit(1)
        
    input_path = Path(args.input)
    if not input_path.exists():
        print(f"Error: Input file {input_path} does not exist.", file=sys.stderr)
        sys.exit(1)
        
    output_path = Path(args.out)
    if output_path.exists() and not args.force:
        print(f"Error: Output file {output_path} already exists. Use --force to overwrite.", file=sys.stderr)
        sys.exit(1)
        
    # Parse key color
    hex_color = args.key.lstrip("#")
    if len(hex_color) != 6:
        print("Error: Key color must be a 6-character hex string (e.g. #00ff00)", file=sys.stderr)
        sys.exit(1)
    key_r = int(hex_color[0:2], 16)
    key_g = int(hex_color[2:4], 16)
    key_b = int(hex_color[4:6], 16)
    
    img = Image.open(input_path).convert("RGBA")
    datas = img.getdata()
    
    new_data = []
    tolerance = args.tolerance
    
    for item in datas:
        r, g, b, a = item
        # Calculate Euclidean distance in RGB space
        dist = ((r - key_r) ** 2 + (g - key_g) ** 2 + (b - key_b) ** 2) ** 0.5
        if dist < tolerance:
            # Fully transparent
            new_data.append((r, g, b, 0))
        elif dist < tolerance + 30:
            # Soft edge transition
            factor = (dist - tolerance) / 30.0
            new_alpha = int(255 * factor)
            new_data.append((r, g, b, new_alpha))
        else:
            new_data.append((r, g, b, a))
            
    img.putdata(new_data)
    img.save(output_path, "PNG")
    print(f"Chroma key background removed. Saved output to {output_path}")

if __name__ == "__main__":
    main()
